org.eclipse.buildship.core/src/main/java/org/eclipse/buildship/core/workspace/internal/GradleClasspathContainerUpdater.java

```diff
@@ -80,39 +81,64 @@ private void updateClasspathContainer(IProgressMonitor monitor) throws JavaModel
-                }).toList();
-
-        // external dependencies
-        List<IClasspathEntry> externalDependencies = FluentIterable.from(this.gradleProject.getExternalDependencies()).filter(new Predicate<OmniExternalDependency>() {
+        boolean hasExportedEntry = FluentIterable.from(externalDependencies).anyMatch(new Predicate<IClasspathEntry>() {
 
             @Override
-            public boolean apply(OmniExternalDependency dependency) {
-                File file = dependency.getFile();
-                String name = file.getName();
-                // Eclipse only accepts folders and archives as external dependencies (but not, for example, a DLL)
-                return file.isDirectory() || name.endsWith(".jar") || name.endsWith(".zip");
+            public boolean apply(IClasspathEntry entry) {
+                return entry.isExported();
             }
-        }).transform(new Function<OmniExternalDependency, IClasspathEntry>() {
+        });
+
+        // Gradle distributions <2.5 rely on exports to define the project classpath. Unfortunately
+        // that logic is broken if dependency excludes are defined in the build scripts. To work
+        // around that, external dependencies must be defined before project dependencies. For more
+        // details, visit Bug 473348.
+        if (hasExportedEntry) {
+            return ImmutableList.<IClasspathEntry>builder().addAll(externalDependencies).addAll(projectDependencies).build();
+        } else {
+            return ImmutableList.<IClasspathEntry>builder().addAll(projectDependencies).addAll(externalDependencies).build();
+        }
+    }
 
-            @Override
-            public IClasspathEntry apply(OmniExternalDependency dependency) {
-                IPath file = org.eclipse.core.runtime.Path.fromOSString(dependency.getFile().getAbsolutePath());
-                IPath sources = dependency.getSource() != null ? org.eclipse.core.runtime.Path.fromOSString(dependency.getSource().getAbsolutePath()) : null;
-                return JavaCore.newLibraryEntry(file, sources, null, ClasspathUtils.createAccessRules(dependency), ClasspathUtils.createClasspathAttributes(dependency), dependency.isExported());
+    private List<IClasspathEntry> collectExternalDependencies() {
+        Builder<IClasspathEntry> result = ImmutableList.builder();
+        for (OmniExternalDependency dependency : this.gradleProject.getExternalDependencies()) {
+            File dependencyFile = dependency.getFile();
+            String dependencyName = dependencyFile.getName();
+            if (!dependencyFile.exists()) {
+                IPath path = new Path("/" + dependencyFile.getPath());
+                IResource member = this.eclipseProject.getProject().findMember(path);
+                if (member != null) {
+                    IClasspathEntry entry = JavaCore.newLibraryEntry(member.getFullPath(), null, null);
+                    result.add(entry);
+                    continue;
```

Comment:
You mean `if (!dependencyFile.exists()) { ... } else { /* add classic deps */ }`?
That way we wouldn't declare those dependencies that are deleted from the Gradle cache. I'm referring your comment here: https://github.com/gradle/tooling-commons/commit/5eddc913cf92392d8bb77d14878f2ca2a7f861f9#commitcomment-20713851